### DISCUSSION TITLE
Using lazy-fixture with parametrize and sharing the dataset

### DISCUSSION THREAD

**Discussion by janemba** [pytest-dev/pytest#13077] [Q&A] [answered]
Hello,

I would like to use `lazy-fixture` and `mark.parametrize` with lists.

My case is a raw list with a dictionary and a generated dataclass list. What I want to achieve is comparing the raw list against the dataclass generated list.

The dataclass is represented below:
```
@dataclasses.dataclass
class Personal:
    id: int
    name: str

@dataclasses.dataclass
class Personals:
    persons: list[Personal]

    def add(self, identifier: int, name: str) -> None:
        self.persons.append(Personal(identifier, name))
```

As an example, I used the following `conftest.py` file:
```
@pytest.fixture(scope="session")
def raw_data():
    with open(filepath, "r") as rd:
        data = rd.readlines()

    raw = list()
    for line in data:
        csv = line.split(",")
        raw.append({"id": int(csv[0]), "name": csv[1]})

    return raw

@pytest.fixture(scope="session")
def generated_data(raw_data):
    personals = Personals()

    for data in raw_data:

        personals.add(data['id'], data['name'])

    return personals

@pytest.fixture(scope="session")
def reuse_generated_data(generated_data):

    return generated_data.persons

@pytest.fixture(scope="session")
def reuse_combined_data(reuse_generated_data, raw_data):
    return list(zip(reuse_generated_data, raw_data))
```

So now, I do the following test in "test_filling.py":
```
@pytest.mark.parametrize("generated, raw", pytest.lazy_fixture("combined_data"))
def test_compare_id(generated, raw):

    assert generated.id == raw["id"], f"Mismatch in 'id': {generated.id} != {raw['id']}"
```

I ended-up with the following error: `TypeError: 'LazyFixture' object is not iterable`

How can I fix this ?

**Comment by kwist-sgr** [comment]
According to the [documentation](https://github.com/TvoroG/pytest-lazy-fixture/tree/master?tab=readme-ov-file#usage)
```
pytest-lazy-fixture lets you use a fixture as one of the values passed in @pytest.mark.parametrize
```

```python
import pytest
from pytest_lazyfixture import lazy_fixture

@pytest.fixture
def one():
    return 1

@pytest.mark.parametrize('arg1,arg2', [
    ('val1', lazy_fixture('one')),
])
def test_func(arg1, arg2):
    assert arg2 == 1
```

See also: https://github.com/TvoroG/pytest-lazy-fixture/issues/56

**Reply by janemba** [reply]
That is not exactly my case.

My test cases are in files. So from this file I'm getting the raw data with the expected values. Also, those data goes into the dataclasses for processing and this is what I want to test.

So the generated list and the raw list cannot be hard-coded and that's why the generated list (the one to test against) and the raw list (the one with expected values) are into functions.

In the example you shown and also in the linked thread, only hard-coded values are used with `params`. In my case, this is a list coming from a `function`. Can `params` be used with functions ? No that's not possible.

**Comment by kwist-sgr** [comment]
I guess it should look something like this:

```python
@pytest.mark.parametrize("generated, raw", [
    (
        pytest.lazy_fixture("fixture_generated"), 
        pytest.lazy_fixture("fixture_raw")),
    )
])
def test_compare_id(generated, raw):
    ...
```

**Reply by janemba** [reply]
I tried your example and I'm having a different error and it seems to be linked to lazy-fixture.

```
  Plugin: lazy-fixture, Hook: pytest_generate_tests
  AttributeError: 'CallSpec2' object has no attribute 'funcargs'
  For more information see https://pluggy.readthedocs.io/en/stable/api_reference.html#pluggy.PluggyTeardownRaisedWarning
    self.ihook.pytest_generate_tests.call_extra(methods, dict(metafunc=metafunc))

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
================================================================================== short test summary info ===================================================================================
ERROR tests/test_filling.py - AttributeError: 'CallSpec2' object has no attribute 'funcargs'
```

I read the URL from the error but it doesn't help. Apparently, the issue is related to the implementation of the plugin.

**Reply by kwist-sgr** [reply, marked as answer]
https://github.com/TvoroG/pytest-lazy-fixture/issues/65

**Reply by janemba** [reply]
Thanks to your link, I removed `pytest-lazy-fixture` and use `request` module. It works ! But, now I have to iterate on the list in my test function.
```python
@pytest.mark.parametrize("generated, raw", [
    (
        "reuse_generated_data",
        "raw_data"
    )
])
def test_compare_id(generated, raw, request):

    generated = request.getfixturevalue(generated)
    raw = request.getfixturevalue(raw)

    for index, expected in enumerate(raw):
        assert generated[index].id == expected["id"], f"Mismatch in 'id' at index {index}"
```

I don't know if it is a good way to do it but is it possible to avoid the iteration and having `parametrize` to give each items of the lists instead ?

**Reply by kwist-sgr** [reply]
Unfortunately, no.

But maybe this option will fit:

```python
@pytest.fixture
def combine_data(reuse_generated_data, raw_data):
    return enumerate(zip(reuse_generated_data, raw_data))

def test_compare_id(combine_data):
    for index, (person, expected) in combine_data:
        assert person.id == expected["id"], f"Mismatch in 'id' at index {index}"
```

**Reply by janemba** [reply]
Yes this works.

Finally, I found a solution. In fact, I don't need two functions to generate the expected data and the generated ones. I do it in one function and then I return a tuple. The function is a generator and I remove the fixture.

Basically `conftest.py` look like this:
```python
def  generated_raw_data():
    with open(filepath, "r") as rd:
        data = rd.readlines()

    personals = Personals()
    raw = list()
    for line in data:
        csv = line.split(",")
        raw.append({"id": int(csv[0]), "name": csv[1]})
        personals.add(id, name)

    yield (personals.persons, raw)

@pytest.fixture(scope="session", params = generated_raw_data())
def given_expected(request):

    return request.param
```

Then, I'm performing my tests as follow:
```python
class TestFilling:

    def test_id(self, given_expected):

        generated, raw = given_expected

        assert generated.id == raw["id"], f"Mismatch in 'id': {generated.id} != {raw['id']}"
```

The only thing is that I'm not using `parametrize`, I don't know the implication except that I can't set `ids`.

Thank you for your help.
